src/main/java/hudson/remoting/Response.java

```diff
@@ -69,23 +77,58 @@
      * Notifies the waiting {@link Request}.
      */
     @Override
-    protected void execute(Channel channel) {
+    void execute(Channel channel) {
         Request req = channel.pendingCalls.get(id);
         if(req==null)
             return; // maybe aborted
         req.responseIoId = lastIoId;
 
         req.onCompleted(this);
         channel.pendingCalls.remove(id);
+        request = req;
+        long startTime = req.startTime;
+        if (startTime != 0) {
+            long time = System.nanoTime() - startTime;
+            totalTime = time;
+            channel.notifyResponse(req, this, time);
+        }
     }
 
+    @Override
     public String toString() {
-        return "Response[retVal="+toString(returnValue)+",exception="+toString(exception)+"]";
+        return "Response:" + request + "(" + (returnValue != null ? returnValue.getClass().getName() : exception != null ? exception.getClass().getName() : null) + ")";
```

Comment:
Will generate a barely comprehensible output if the request is null. Maybe makes sense to do `Response:
 [request=..., retval=...]` or so.